File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    visited = maps
    lenX = len(maps);
    lenY = len(maps[0]);
    dx = [-1, 1, 0, 0];
    dy = [0, 0, -1, 1];
    
    queue = deque();
    queue.append([0, 0]);
    
    while queue:
        x, y = queue.popleft();
        
        for i in range(4):
            nextX, nextY = x + dx[i], y + dy[i];
            if 0 <= nextX < lenX and 0 <= nextY < lenY and visited[nextX][nextY] == 1:
                visited[nextX][nextY] = visited[x][y] + 1;
                queue.append([nextX, nextY]);
                
    return visited[-1][-1] if visited[-1][-1] > 1 else -1

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const visited = [...maps];
    const lenX = maps.length;
    const lenY = maps[0].length;
    const dx = [-1, 1, 0 ,0];
    const dy = [0, 0, -1, 1];
    
    const queue = [[0, 0]];
    
    while(queue.length){
        let [x, y] = queue.shift();
        for(let i = 0; i < 4; i++){
            let nextX = x + dx[i], nextY = y + dy[i];
            if(nextX >= 0 && nextX < lenX 
               && nextY >= 0 && nextY < lenY 
               && visited[nextX][nextY] === 1){
                    visited[nextX][nextY] = visited[x][y] + 1;
                    queue.push([nextX, nextY])
            }
        }
    }
    return visited[lenX - 1][lenY - 1] > 1 ? visited[lenX - 1][lenY -1] : -1
}